### PR TITLE
Added improvements, better error handling, max queue size to prevent memory runaway, etc.

### PR DESCRIPTION
**Pull request by keithjjones**
Thanks for making a native Python library!  

I had two issues using it, so I took the liberty of creating fixes and this PR.  The first issue was with stopping of the sub processes.  The processes would often hang.  It worked out better to terminate the processes.  Second, the memory would runaway for very active streams.  This would often kill the docker container it was running in.  I added a max size to the bigger queues so that the overall memory could be reduced.  The native functionality is still there if you leave max size to None.  Also, I added a lot more logging in the more critical areas that were causing errors for me.

I tried pytest, but it was looking for "mocker", which I didn't see in the code.  Therefore, I don't know if the tests pass, but they should because I'm using this code successfully on a Kinesis stream right now.

Thanks!

### CODE REVIEW

**Comment by borgstrom** [comment]
Thanks for the PR @keithjjones!

I'll do a full review on this in the next couple days, just wanted to let you know it's in queue.

**Comment by mwarkentin** [comment]
Any updates on this? This PR looks like it answers a couple of questions we had when initially reviewing this library for use as we start to use kinesis. :)

**Comment by hampsterx** [comment]
sorry to +1 but this would be great to get merged in!~ @borgstrom

**Comment by hampsterx** [comment]
ok just tested this out and seems awesome!

readme needs to be updated to state the producer needs to use start() and shutdown() afterwards.
Also on linux/python3.6 I get 

```
AttributeError: 'Process' object has no attribute 'kill'  at Producer.py line 193
```
Probably 

```
self.async_producer.kill()
```

should instead be ?

```
self.async_producer.terminate()
```

**Comment by hampsterx** [comment]
@borgstrom sorry to nag but any thoughts on this? would be great to get merged in so this project could move forward :)

File: src/kinesis/consumer.py [hunk 1]
```diff
@@ -1,234 +1,341 @@
...
+    def __iter__(self):
+        self.start()
+        return self
+
+    def __next__(self):
+        return self.get_record()
+
+    def start(self):
+        log.debug("Starting Consumers")
+        if hasattr(self, "manager_proc") and (self.manager_proc is None or
```
Comment (hampsterx) [inline src/kinesis/consumer.py:57]: > hasattr(self, "manager_proc") 

property is declared as None so this is not necessary.

File: src/kinesis/consumer.py [hunk 2]
```diff
@@ -1,234 +1,341 @@
...
-            # use lock duration - 1 here since we want to renew our lock before it expires
-            lock_duration_check = self.LOCK_DURATION - 1
-            while self.run:
-                last_setup_check = time.time()
-                self.setup_shards()
-
-                while self.run and (time.time() - last_setup_check) < lock_duration_check:
+        log.info("All shard readers have been shutdown.")
+        log.info("Shutting down shard manager...")
+        if hasattr(self, "manager_proc"):
```
Comment (hampsterx) [inline src/kinesis/consumer.py:247]:
```
if self.manager_proc
```

this will ensure if start_shard_manager() not called it won't bomb out here.

**Comment by hampsterx** [comment]
@keithjjones pretty please can you rebase/fix the conflict :)

**Comment by keithjjones** [comment]
Hi I haven't touched this in quite a while.  Can you work with the code I submitted?

**Comment by hampsterx** [comment]
hi @keithjjones I am just starting on using it for a project now (got delayed for a while), just getting beyond the basics, seems to work so far. I guess I could fork your fork (actually I already did) and submit a new PR to fix the conflicts and apply any other tweaks?

Update: never mind, wrote my own lib https://github.com/hampsterx/async-kinesis

**Comment by keithjjones** [comment]
No complaints out of me.  I won't be able to get to it for at least a month, I'm figuring.

**Comment by borgstrom** [comment]
@keithjjones -- Can you share some info on why you choose to use multiprocessing directly instead of the offspring library?

This PR is very large and hard to review because there are so many different changes all lumped into a same change set.

I'm very open to incorporating some of your improvements, but the changes need to be done in smaller pieces so that they are easier to review.

**Pull request closed without merging**